### PR TITLE
Raise retry number to 5 and activate dead lettering

### PR DESCRIPTION
**Pull request by amiedes**
Story details: https://app.clubhouse.io/cartoteam/story/128622

## What does this PR do

- Allows specifying a Dead Letter policy when creating a subscription
- Modifies the mesage-handling logic so when an error occurs while processing it, we do a `nack!` instead of an `ack!` so the message is available for redelivery
- Exposes the Pub/Sub project name and Subscriptions full name (with the preceing project ID etc) in order to grant the necessary Dead Letter permissions from Central

### CODE REVIEW

File: cartodb-utils.gemspec [hunk 1]
```diff
@@ -19,9 +19,10 @@ Gem::Specification.new do |spec|
     `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^spec/}) }
   end
 
-  spec.add_dependency 'rails', '>= 4', '< 6'
   spec.add_dependency 'argon2', '~> 2'
   spec.add_dependency 'google-cloud-pubsub', '~> 2.3'
+  spec.add_dependency 'google-cloud-resource_manager'
```
Comment (amiedes) [inline cartodb-utils.gemspec:24]: This dependency was needed in order to retrieve the Pub/Sub project number. In the `app_config.yml` we have the text-like ID, but I needed the number to setup the necessary permissions in https://github.com/CartoDB/cartodb-central/pull/3025/files#diff-c48f1047908bcbf32048b8ed8da97ecd68084698506523a41ee8b9d036cf06ed

File: lib/carto/common/message_broker.rb [hunk 2]
```diff
@@ -50,7 +53,8 @@ def get_topic(topic)
       def create_topic(topic)
         topic_name = pubsub_prefixed_name(topic)
         begin
-          @pubsub.create_topic(topic_name)
+          new_topic = @pubsub.create_topic(topic_name)
+          logger.info(message: 'Topic created', topic_name: new_topic.name)
```
Comment (amiedes) [inline lib/carto/common/message_broker.rb:57]: I've added some minor log traces, as it turned out useful while debugging some container crashes during the environment seeding

File: lib/carto/common/message_broker.rb [hunk 3]
```diff
@@ -108,7 +124,7 @@ def initialize(pubsub, project_id:, topic_name:, logger: nil, publisher_validati
           @pubsub = pubsub
           @project_id = project_id
           @topic_name = topic_name
-          @topic = @pubsub.get_topic("projects/#{@project_id}/topics/#{@topic_name}")
+          @topic = get_topic(@topic_name)
```
Comment (amiedes) [inline lib/carto/common/message_broker.rb:127]: Just extracting the `get_topic` logic into a method, as I was going to reuse it in another place

File: lib/carto/common/message_broker.rb [hunk 4]
```diff
@@ -125,12 +141,19 @@ def publish(event, payload)
...
-            subscription_name = pubsub_prefixed_name(subscription)
-            @topic.create_subscription(subscription_name,
-                                       deadline: SUBSCRIPTION_ACK_DEADLINE_SECONDS,
-                                       retry_policy: SUBSCRIPTION_RETRY_POLICY)
+            subscription_name = pubsub_prefixed_name(subscription_name)
+            subscription = @topic.create_subscription(
+              subscription_name,
+              deadline: ack_deadline_seconds,
+              retry_policy: retry_policy,
+              dead_letter_topic: dead_letter_topic_name ? get_topic(pubsub_prefixed_name(dead_letter_topic_name)) : nil,
```
Comment (amiedes) [inline lib/carto/common/message_broker.rb:153]: If a `dead_letter_topic_name` is provided, find the corresponding Pub/Sub topic and propagate it as an argument to `Google::Cloud::PubSub::Topic#create_subscription` so it sets up dead lettering

File: lib/carto/common/message_broker.rb [hunk 5]
```diff
@@ -125,12 +141,19 @@ def publish(event, payload)
           result
         end
 
-        def create_subscription(subscription)
+        def create_subscription(subscription_name, ack_deadline_seconds: DEFAULT_SUBSCRIPTION_ACK_DEADLINE_SECONDS,
+                                dead_letter_topic_name: nil, dead_letter_max_delivery_attempts: nil,
+                                retry_policy: DEFAULT_SUBSCRIPTION_RETRY_POLICY)
```
Comment (amiedes) [inline lib/carto/common/message_broker.rb:146]: Modify the `Carto::Common::MessageBroker::Topic#create_subscription` API so it accepts customizing the retry policy and the dead letter policy

File: lib/carto/common/message_broker.rb [hunk 6]
```diff
@@ -239,6 +269,9 @@ def main_callback(received_message)
         def start(options = {})
           logger.info(message: 'Starting message processing in subscriber',
                       subscription_name: @subscription_name)
+
+          raise NotFound, "Subscription #{@subscription_name} does not exist" if @subscription.blank?
```
Comment (amiedes) [inline lib/carto/common/message_broker.rb:273]: The purpose was just to be able to see the name of the missing subscription in development if some error happens during the containers boot

File: lib/carto/common/message_broker.rb [hunk 7]
```diff
@@ -226,7 +255,8 @@ def main_callback(received_message)
                            },
                            subscription_name: @subscription_name,
                            message_type: message_type)
-              received_message.ack!
+              # Make the message available for redelivery
+              received_message.reject!
```
Comment (amiedes) [inline lib/carto/common/message_broker.rb:259]: ⚠️ ⚠️ ⚠️

File: spec/carto/common/message_broker/subscription_spec.rb [hunk 8]
```diff
@@ -2,12 +2,26 @@
...
+  before do
+    allow(pubsub).to receive(:get_subscription).with('test_subscription', project: 'test-project-id')
+    allow(message).to receive(:ack!)
+    allow(message).to receive(:attributes).and_return({ 'event' => 'dummy_command' })
+    allow(message).to receive(:data).and_return('{}')
+  end
 
   describe '#initialize' do
     it 'delegates on the pubsub object to get the subscription' do
-      pubsub = instance_double('PubsubDouble')
```
Comment (amiedes) [inline spec/carto/common/message_broker/subscription_spec.rb:8]: Just extracting some common setup between the specs in this file

Comment (rafatower) [reply spec/carto/common/message_broker/subscription_spec.rb:8]: thanks!

File: lib/carto/common/message_broker.rb [hunk 9]
```diff
@@ -19,9 +20,11 @@ def pubsub_prefixed_name(input_symbol_or_string)
 
     class MessageBroker
 
-      SUBSCRIPTION_ACK_DEADLINE_SECONDS = 300
-      SUBSCRIPTION_RETRY_POLICY = Google::Cloud::PubSub::RetryPolicy.new(minimum_backoff: 10,
-                                                                         maximum_backoff: 600)
+      DEFAULT_SUBSCRIPTION_ACK_DEADLINE_SECONDS = 300
+      DEFAULT_SUBSCRIPTION_RETRY_POLICY = Google::Cloud::PubSub::RetryPolicy.new(minimum_backoff: 10,
+                                                                                 maximum_backoff: 600)
+      DEFAULT_DEAD_LETTER_TOPIC_NAME = :dead_letter_queue
```
Comment (rafatower) [inline lib/carto/common/message_broker.rb:26]: Minor: I think this `DEFAULT_DEAD_LETTER_TOPIC_NAME` is not used.

**Review by rafatower — APPROVED** [review]

File: lib/carto/common/message_broker.rb (hunk 9 above)
Comment (amiedes) [reply lib/carto/common/message_broker.rb:26]: It's used a couple of times but from Central, not from this PR itself (forgot to mention)

**Pull request merged**